### PR TITLE
feat/support reachable vulns

### PR DESCRIPTION
**Pull request by darmalovan**
- [x] Ready for review
- [x] Follows CONTRIBUTING rules
- [ ] Reviewed by Snyk internal team

#### What does this PR do?
If the `--reachable-vulns` flag is being passed from the CLI,
the plugin will create a functions call graph and pass it
back to the CLI.

The way the graph is The java call graph builder generated
is by downloading a jar that will produce the call graph.
The jar is only downloaded if needed, more about this logic
is in https://github.com/snyk/java-call-graph-builder

Also, add node 12 to the test matrix.

#### Any background context you want to provide?
The tests are failing due the dependency (java-call-graph-builder) not supporting node 8, working on this in the meanwhile.

#### What are the relevant tickets?
https://snyksec.atlassian.net/browse/FLOW-159

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -39,7 +39,8 @@
     "tap-only": "0.0.5"
   },
   "dependencies": {
-    "@snyk/cli-interface": "2.3.1",
+    "@snyk/cli-interface": "2.4.0",
+    "@snyk/java-call-graph-builder": "^1.3.1",
```
Comment (darmalovan) [inline package.json:43]: This will be updated to support node 8

File: .travis.yml [hunk 2]
```diff
@@ -39,6 +39,14 @@ jobs:
         - sdk install maven $MAVEN_VERSION
         - java -version
         - npm test
+    - node_js: 12
```
Comment (darmalovan) [inline .travis.yml:42]: Adding support to node 12

File: tests/system/get-command.test.ts [hunk 3]
```diff
@@ -1,5 +1,5 @@
 import * as path from 'path';
-import { test, Test } from 'tap';
```
Comment (darmalovan) [inline tests/system/get-command.test.ts:2]: Was unused

File: lib/index.ts [hunk 4]
```diff
@@ -3,11 +3,18 @@ import * as fs from 'fs';
...
+import * as javaCallGraphBuilder from '@snyk/java-call-graph-builder';
 import { containsJar, createPomForJar, createPomForJars, isJar } from './jar';
 import * as os from 'os';
 
+import debugLib = require('debug');
+const debug = debugLib('snyk-mvn-plugin');
+
 export interface MavenOptions extends legacyPlugin.BaseInspectOptions {
   scanAllUnmanaged?: boolean;
+  'reachable-vulns'?: boolean;
```
Comment (dkontorovskyy) [inline lib/index.ts:17]: How about camel case this is cli?

File: tests/fixtures/test-project/expected-with-call-graph.json [hunk 5]
```diff
@@ -0,0 +1,48 @@
+{
+  "plugin": { "name": "bundled:maven", "runtime": "unknown" },
+  "callGraph": {},
```
Comment (dkontorovskyy) [inline tests/fixtures/test-project/expected-with-call-graph.json:3]: Will there be a test with actual or semi-actual call graph?

File: lib/index.ts [hunk 6]
```diff
@@ -3,11 +3,18 @@ import * as fs from 'fs';
 import * as path from 'path';
 import * as subProcess from './sub-process';
 import { legacyPlugin } from '@snyk/cli-interface';
+import { CallGraph } from '@snyk/cli-interface/legacy/common';
+
```
Comment (dkontorovskyy) [inline lib/index.ts]: Please remove this empty line 🙏

File: lib/index.ts (hunk 4 above)
Comment (darmalovan) [reply lib/index.ts:17]: I tried to do something like that in the CLI, but it's pretty ugly. The reason is that in the CLI the `Options & TestOptions` structure is being passed to the plugin, by having the `reachable-vulns` flag set up from the user, I would need to have `reachable-vulns` and `reachableVulns` at the same object.

File: tests/system/plugin.test.ts [hunk 7]
```diff
@@ -4,7 +4,11 @@ import * as plugin from '../../lib';
 import { readFixtureJSON } from '../file-helper';
 import { legacyPlugin } from '@snyk/cli-interface';
 import * as os from 'os';
+import * as javaCallGraphBuilder from '@snyk/java-call-graph-builder';
+import sinon = require('sinon');
```
Comment (dkontorovskyy) [inline tests/system/plugin.test.ts:8]: Maube add `@types/sinon` to not have `require`

File: tests/fixtures/test-project/expected-with-call-graph.json (hunk 5 above)
Comment (darmalovan) [reply tests/fixtures/test-project/expected-with-call-graph.json:3]: It's pretty heavy to test to full functionality of the call graph builder, I can maybe add a more useful graph just for the fixtures.

Comment (darmalovan) [reply tests/fixtures/test-project/expected-with-call-graph.json:3]: A full integration test will be part of the CLI

**Comment by darmalovan** [comment]
@dkontorovskyy I've addressed your comments and raised another commit.

File: lib/index.ts (hunk 4 above)
Comment (dkontorovskyy) [reply lib/index.ts:17]: But then you wouldn't have this in code anywhere `'reachable-vulns'`

Comment (darmalovan) [reply lib/index.ts:17]: In the CLI I'll have it

Comment (darmalovan) [reply lib/index.ts]: @dkontorovskyy managed to find the mechanism that removes dashes from the CLI! updated the latest commit

**Pull request merged**